### PR TITLE
RC2 Breaking - Ensure migrations persist the executed key, when executed.

### PR DESCRIPTION
**Pull request by bergmania**
### Description
Adds new functionality to the migrations, to avoid the situration where a migration is executed, but Umbraco stops, before the KeyValue with migration state is updated.

This requires a migration to call `Context.Complete()` on the migration context. This happens automatically on scoped migrations before the scope is completed. But migrations inheriting the `UnscopedMigrationBase` needs to call this manually, inside the scopes or at least when it is considered done.

Thereby, we minimize the risk (and eliminate it for SqlServer) that a migration is executed but the state is not saved.

If a migration is executed without the `Complete` is called, the migration upgrader throws an error, so we do not start executing the next migration

### Breaking changes
- Migrations that inherit `UnscopedMigrationBase` now have to call `Context.Complete()` when done.

### Test
- Ensure migrations from v13 is possible
  - [ ] Sqlite
  - [ ] SqlServer

### CODE REVIEW

File: src/Umbraco.Infrastructure/Migrations/IMigrationContext.cs [hunk 1]
```diff
@@ -44,4 +44,8 @@ public interface IMigrationContext
     [Obsolete("This will be removed in the V13, and replaced with a RebuildCache flag on the MigrationBase")]
     void AddPostMigration<TMigration>()
         where TMigration : MigrationBase;
+
+    public bool IsDone { get; }
+
+    public void SetDone();
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Migrations/IMigrationContext.cs:50]:
```suggestion

    public bool IsCompleted { get; }

    public void Complete();
```

File: src/Umbraco.Infrastructure/Migrations/MigrationContext.cs [hunk 2]
```diff
@@ -8,13 +10,15 @@ namespace Umbraco.Cms.Infrastructure.Migrations;
...
 {
+    private readonly Action _onDoneAction;
     private readonly List<Type> _postMigrations = new();
 
     /// <summary>
     ///     Initializes a new instance of the <see cref="MigrationContext" /> class.
     /// </summary>
-    public MigrationContext(MigrationPlan plan, IUmbracoDatabase? database, ILogger<MigrationContext> logger)
+    public MigrationContext(MigrationPlan plan, IUmbracoDatabase? database, ILogger<MigrationContext> logger, Action onDoneAction)
     {
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Migrations/MigrationContext.cs:20]: Can `onDoneAction` be passed in `SetDone` instead, or without this method and call this action in caller only?

File: src/Umbraco.Infrastructure/Migrations/MigrationContext.cs [hunk 3]
```diff
@@ -8,13 +10,15 @@ namespace Umbraco.Cms.Infrastructure.Migrations;
...
+        _onDoneAction = onDoneAction;
```
Comment (Nuklon) [inline src/Umbraco.Infrastructure/Migrations/MigrationContext.cs:21]: This should also be renamed (completeAction?). Same with `OnMigrationDone` method in class below and the comment referring to `OnDone`.

Maybe this last parameter can be optional? Then you don't have to pass in nop action everywhere in tests.

**Review by nikolajlauridsen — APPROVED** [review]
Looks good and tests good 🎉 👍.

I did however remove the `public` access modifiers from the interface, and added a missing space to the error message.

**Pull request merged**